server/worker/apple_mdm.go

```diff
@@ -238,6 +239,13 @@ func (a *AppleMDM) runPostDEPReleaseDevice(ctx context.Context, args appleMDMArg
 		return ctxerr.Wrap(ctx, err, "failed to get host MDM profiles")
 	}
 	for _, prof := range profs {
+		// NOTE: DDM profiles (declarations) are ignored because while a device is
+		// awaiting to be released, it cannot process a DDM session (at least
+		// that's what we noticed during testing).
+		if strings.HasPrefix(prof.ProfileUUID, fleet.MDMAppleDeclarationUUIDPrefix) {
+			continue
+		}
```

Comment:
@georgekarrv So this is a super-quick potential fix: in the job that waits for all pending profiles to be delivered before releasing the device, it will ignore any DDM "profile" (any declaration), so if there's a DDM profile still pending but otherwise all is good, it will send the command to release the device.

I'm not sure if this will reliably work because I think it could end up in a situation where the DDM profile is the next command to run and the DeviceConfigured command is queued after, I'm not totally sure but if you can give this "fix" a try, it could be a first step.

Would be good to also get @roperzh 's input for his knowledge of DDM and nanomdm internals, especially:
a) have you read something official about the fact that a device pending release cannot start a DDM session?
b) could it indeed get in a situation where the DDM command is next in line, receives "NotNow" because the device is not released, and the release command never gets executed (because it is after the DDM profile in the queue)?

I think those are different queues so maybe b) would not happen? But then I'm not sure why we have the issue in the first place, the device should've been released after ~15 minutes in any case.

Reply:
@mna 

>  a) have you read something official about the fact that a device pending release cannot start a DDM session?

The only thing I know is that when a device sends the `DeclarativeManagement` message, (at least mine) the message includes  an `AwaitingConfiguration` key:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>AwaitingConfiguration</key>
	<false/>
	<key>Endpoint</key>
	<string>tokens</string>
	<key>MessageType</key>
	<string>DeclarativeManagement</string>
	<key>UDID</key>
	<string>419D46EC-06E6-557C-AD52-601BA0667730</string>
</dict>
</plist>
```

this makes me think that DDM should work, but I haven't read anything in the docs

>  b) could it indeed get in a situation where the DDM command is next in line, receives "NotNow" because the device is not released, and the release command never gets executed (because it is after the DDM profile in the queue)?

IMO it'll depend on what you saw, if you saw the `DeclarativeManagement` command receive a `NotNow` consistently, then I think you're absolutely right.

If we would like to double down on the quickfix and take a calmer look, we could use the `priority` column after all, to make DDM commands the last items in the queue, wdty?


---

I will try to reproduce on my end. From the logs @georgekarrv posted in the issue, to me it seems like the device acknowledged all the commands, and we run out of commands (see the `Idle` without response at the end)

```
level=info ts=2024-04-09T21:00:10.997468103Z component=http-mdm-apple-mdm id=AF05843A-3C54-5A8F-B248-361A13FD0922 type=Device status=Acknowledged command_uuid=b7bf8148-0f41-4197-b313-13b86fd1b874
level=info ts=2024-04-09T21:00:11.046245573Z component=http-mdm-apple-mdm id=AF05843A-3C54-5A8F-B248-361A13FD0922 type=Device status=Acknowledged command_uuid=7e80a44a-af34-42e8-9fb5-c937c54db697
level=info ts=2024-04-09T21:00:11.993650264Z component=http-mdm-apple-mdm id=AF05843A-3C54-5A8F-B248-361A13FD0922 type=Device status=Acknowledged command_uuid=d1bb3e4d-9ecc-48c6-aa93-0c5417862119
level=info ts=2024-04-09T21:00:12.097029593Z component=http-mdm-apple-mdm id=AF05843A-3C54-5A8F-B248-361A13FD0922 type=Device status=Idle
```

which makes me think the root problem is something else

Reply:
> IMO it'll depend on what you saw, if you saw the DeclarativeManagement command receive a NotNow consistently, then I think you're absolutely right.

Yes that is what we observed in the first instance. I left it for ~1hr and it was just looping on this

Reply:
```
level=error ts=2024-04-09T21:01:10.324062551Z cron=integrations job_id=53 msg="process job" err="running post Apple DEP release device task: device not ready for release, profile passcode-settings is still pending, will retry"
```
This is I think the big problem

Reply:
it works 100% of the time if I remove the ddm profile. Fails 100% of the time (3/3 so far) with the profile

Reply:
gotcha, yeah if that's what you saw then I agree with Martin's hunch that we might still run into this even with this fix.

Reply:
```
	<key>AwaitingConfiguration</key>
	<false/>
```
Also I wonder if we need to use `true` here if the device is still configuring

Reply:
sorry for the back and forth folks, this should work fine after all!

https://github.com/fleetdm/fleet/blob/54e20844d299a94b9ac76945d9ecb5ac34155774/server/mdm/nanomdm/service/nanomdm/service.go#L237-L240

nano will skip to the next command in the queue (see the second argument to `RetrieveNextCommand`

Reply:
@georgekarrv 

> Also I wonder if we need to use true here if the device is still configuring 

sorry for not being clear, that's what the device sends to us

Reply:
> nano will skip to the next command in the queue
@roperzh what I don't understand about this is why the device did not end up being released after ~15 minutes. The `DeviceConfigured` command got enqueued at some point, but for some reason never got executed (IIUC) and the queue got stuck on delivering the DDM profile and receiving NotNows.

Reply:
@mna gotcha, what I understood was that we never sent `DeviceConfigured` because the DDM profile was stuck in pending, we definitely need to dig more